Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -448,54 +452,78 @@
     return curState->curBlendMode;
 }
 
-void CGContextCairo::CGContextShowTextAtPoint(float x, float y, const char* str, DWORD length) {
-    // TODO #924: Implement this with DWrite
-    UNIMPLEMENTED();
+void CGContextCairo::CGContextShowTextAtPoint(float x, float y, const char* str, size_t length) {
+    CGFontRef font = curState->getCurFont();
+    std::vector<CGGlyph> glyphs(length);
+    if (_CGFontGetGlyphsForCharacters(font, str, length, glyphs.data())) {
+        CGContextShowGlyphsAtPoint(x, y, glyphs.data(), length);
+    }
 }
 
-void CGContextCairo::CGContextShowGlyphsAtPoint(float x, float y, WORD* glyphs, int count) {
-    CGSize size;
+void CGContextCairo::CGContextShowGlyphsAtPoint(float x, float y, const CGGlyph* glyphs, size_t count) {
+    CGSize size = CGSizeZero;
 
     _isDirty = true;
 
+    std::vector<int> designUnitAdvances(count);
+    CGFontRef font = curState->getCurFont();
+    CGFontGetGlyphAdvances(font, glyphs, count, designUnitAdvances.data());
+    std::vector<CGSize> advances(count);
+    std::transform(designUnitAdvances.cbegin(),
+                   designUnitAdvances.cend(),
+                   advances.begin(),
+                   [ scale = curState->fontSize / CGFontGetUnitsPerEm(font), &size ](int unscaledAdvance) {
+                       CGFloat advanceWidth = scale * unscaledAdvance;
+                       size.width += advanceWidth;
+                       return CGSize{ advanceWidth, 0 };
+                   });
     switch (curState->textDrawingMode) {
         case kCGTextFill:
         case kCGTextStroke:
         case kCGTextFillStroke:
         case kCGTextFillClip:
         case kCGTextStrokeClip:
         case kCGTextFillStrokeClip:
-            size = CGFontDrawGlyphsToContext(glyphs, count, x, y);
+            curState->curTextPosition = { x, y };
+            CGContextShowGlyphsWithAdvances(glyphs, advances.data(), count);
             break;
 
         case kCGTextClip:
         case kCGTextInvisible:
-            // TODO #924: Update the text position in this case
-            UNIMPLEMENTED();
+            // Do nothing, set text position at end
             break;
     }
 
-    curState->curTextPosition.x = x + size.width;
-    curState->curTextPosition.y = y;
+    curState->curTextPosition = { x + size.width, y };
 }
 
-void CGContextCairo::CGContextShowGlyphsWithAdvances(WORD* glyphs, CGSize* advances, int count) {
-    CGPoint curPos = { curState->curTextPosition.x, curState->curTextPosition.y };
+void CGContextCairo::CGContextShowGlyphsWithAdvances(const CGGlyph* glyphs, const CGSize* advances, size_t count) {
     _isDirty = true;
-
-    for (int i = 0; i < count; i++) {
-        CGFontDrawGlyphsToContext(&glyphs[i], 1, curPos.x, curPos.y);
-        curPos.x += advances[i].width;
-        curPos.y += advances[i].height;
+    CGPoint curPos = curState->curTextPosition;
```

Comment:
>CGPoint curPos = curState->curTextPosition; [](start = 4, length = 43)

doesn't seem to be useful anyway. you can just directly use curTextPosition at the end